[CORE] Recursive descent parser + Core/Client separation

AvitalTamir opened the pull request:
This is a complete refactoring of the core Cyphernetes package.
It includes a from-scratch rewrite of the original goyacc Cyphernetes parser and original lexer in favor of a recursive descent function and a more straightforward lexer implementation.
This removes the dependency on goyacc, makes the parser more maintainable and debuggable, and finally lets us produce a coverage report for the parser and lexer.

Additionally, the core package is now separated to packages `core` (which provides the language parser and relationships engine) and `provider` which is an interface for providing client functionality.
The `apiserver` package contains the original Cyphernetes k8s client.
This new architecture will allow external projects to integrate the Cyphernetes language into their project while bringing in their own K8s client or ever an external DB such as Postgres or Elasticsearch (address issue #146)
- [x] pass all existing parser test cases and operator e2e cases.
- [x] lexer test suite rewrite
- [x] include `pkg` in coverage report

naorpeled reviewed (approved):
LGTM!
great stuff

naorpeled commented:
btw,
I've noticed that when I use `\n` I'm getting `Error >> error parsing query >> parse error: expected MATCH or CREATE, got {0 \}`, we probably should return a more refined error

AvitalTamir commented:
Good catch, will address

AvitalTamir commented on cmd/cyphernetes/main_test.go:
```diff
@@ -102,12 +102,12 @@ func TestCyphernetesShellWithNamespaceFlagHelper(t *testing.T) {
-		checkOutput(t, stderr, "[DEBUG]", "\"cyphernetes shell -l debug\"")
-	}
-}
+// func TestCyphernetesShellWithLogLevelFlag(t *testing.T) {
```
Oh yes good catch, not directly related to the PR change. this was a flakey before in CI and failed for me locally when working on the feature. I think a main.go testing issue we already encountered.
Will verify if this is necessary and will either uncomment, remove or reorder tests.

AvitalTamir replied:
Fixed, uncommented back.

AvitalTamir commented:
@naorpeled I addressed both issues. Test is not uncommented and the error messages improved.
BTW I believe the \n without namespace name was a parsing error in the previous version too, according to the code I fixed :)

AvitalTamir commented:
BTW your comment also helped me improve the parsing error, changed it from this:
`Error >> error parsing query >> parse error: expected MATCH or CREATE, got {0 \}`

To this:
`Error >> error parsing query >> parse error: expected MATCH or CREATE, got "\"`

AvitalTamir commented:
@naorpeled This baby is ready 😂 
I think I hashed out every single little issue and this is now fully refactored, complete with engine replacement.
I tested the crap out of it myself, added test cases... Tomorrow I will sit and test this again myself thoroughly.
If you can find the time today/tomorrow - let's get together to go over the PR, maybe you can play with it a little, try to break it and see if there's anything I missed.

AvitalTamir commented:
To whom it may concern,
This PR is the work of a madman who failed to recognize his own mortality and labored for almost 72h straight over refactoring essentially every single file in an entire monorepo spanning 3 different software packages.
This work has NOT been made possible by AI, which apparently, at the time of writing, is a horrible tool for the job.
I initially thought this would be a 3-4 file job, up until the point I realized the fundamental `FindGVR` function did not belong to the `QueryExecutor` (K8s client) structure I was refactoring. This quickly turned into one of those situations like when you have a tiny bit of loose skin next to your fingernail that you start pulling on and pretty soon you realize you've undone your entire finger, it bleeds pretty bad, you're in a great deal of pain you did NOT see coming and likely need to see a medical professional.

elliotxx commented:
@AvitalTamir Brother, I noticed that you have accomplished such a feat. It's so cool, it makes it easier for me to integrate [Cyphernetes](https://github.com/AvitalTamir/cyphernetes) in [Karpor](https://github.com/kusionStack/karpor)!

elliotxx commented:
So far, I am iterating the functionality of v0.6.0 and will need some more time to start integrating Cyphernets...

AvitalTamir commented:
Hey @elliotxx I was about to reach out to you - I'm about to merge another PR which cleans up this interface further, and includes an integration guide :)
I'll tag you when I do.
If you'd like, we can set up some time to go over this and see if there's anything I can do to help further.

AvitalTamir commented:
@elliotxx the interface cleanup is now merged - check out the [integration guide](https://github.com/AvitalTamir/cyphernetes/blob/main/docs/INTEGRATION.md) and lmk if there's anything we can do to make this simpler - I'm keen on making this as frictionless as possible.

elliotxx commented:
@AvitalTamir OK, let me study first. If there is a problem, I will come to you~
